### PR TITLE
Reader: Make the excerpts we use for related posts better

### PR DESCRIPTION
**Pull request by blowery**
And remove some cruft from jetpack sites. Especially LaughingSquid.

This teaches the post normalizer a new rule: `sanitizeContent`. It removes a bunch of cruft from many posts that use common WordPress plugins, like Jetpack Related Posts, WordAds, Google Ads, form elements, etc. 

To test, try http://calypso.localhost:3000/read/feeds/12098 and click through into some posts, then click on the related post links. Notice most of the cruft at the top and bottom of each post is missing in full post view, and the excerpts are much better.

cc @beaulebens as this would be much better to fix on ingest, rather than pulling it out in the client.

### CODE REVIEW

**Comment by fraying** [comment]
Looks like a great improvement!

On this page ... https://calypso.live/read/blogs/1065188/posts/526669 ... I see "[jetpack-related-posts]" at the bottom. Could we yank that, too?

**Comment by fraying** [comment]
Also on this page ... https://calypso.live/read/blogs/1065188/posts/526669 ... note that the post is "German Artist Stefan Pabst Creates a Beautiful 3D Optical Illusion Drawing of the Eiffel Tower" and the two "More in Site" links are also "German Artist Stefan Pabst Creates a Beautiful 3D Optical Illusion Drawing of the Eiffel Tower". They're coming up as different URLs. Any idea why this is happening?

screen shot 2016-10-28 at 11 33 36 am

File: client/lib/post-normalizer/rule-content-sanitize.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+].join( ', ' );
+
+export default function sanitizeContent( post, dom ) {
+	if ( ! dom ) {
+		throw new Error( 'this transform must be used as part of withContentDOM' );
+	}
+
+	const elements = dom.querySelectorAll( thingsToHide );
+	forEach( elements, function( element ) {
+		element.parentNode && element.parentNode.removeChild( element );
```
Comment (samouri) [inline client/lib/post-normalizer/rule-content-sanitize.js]: is there any reason you can't do `element.remove()`?

File: client/lib/post-normalizer/rule-content-sanitize.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+ * Internal Dependencies
+ */
+
+const thingsToHide = [
+	'.sharedaddy', // share daddy
+	'script', // might be too late for these at this point...
+	'.jp-relatedposts', // jetpack related posts
+	'.mc4wp-form', // mailchimp 4 wp
+	'.wpcnt', // wordads?
+	'.OUTBRAIN',
```
Comment (samouri) [inline client/lib/post-normalizer/rule-content-sanitize.js]: what is outbrain?

File: client/lib/post-normalizer/rule-content-sanitize.js (hunk 1 above)
Comment (blowery) [inline client/lib/post-normalizer/rule-content-sanitize.js]: only works on Edge, not IE

File: client/lib/post-normalizer/rule-content-sanitize.js (hunk 2 above)
Comment (blowery) [inline client/lib/post-normalizer/rule-content-sanitize.js]: a content ad system www.outbrain.com

**Comment by blowery** [comment]
@gibrown @aaronyan any thoughts on the source post showing up in related posts from the same site?

**Comment by aaronyan** [comment]
Can a site reblog their own post? If so, there is a chance that the reblogged post made it into the "More in Laughing Squid" section.

@fraying, @blowery: Is this still an issue? I'm having trouble reproducing this:

laughing_squid_german_artist_stefan

@gibrown: I think we have a bug where reblogged posts are not being indexed correctly in Elasticsearch. I've seen [reblogged posts appear in related post results](https://wordpress.com/read/feeds/41325786/posts/1171289364). We'll chat about making this a project milestone.

**Comment by gibrown** [comment]
@aaronyan let's just filter out all reblogs. There is a boolean field called `is_reblogged`

**Comment by aaronyan** [comment]
@gibrown: sorry for not being clear. `is_reblogged` doesn't seem to be working. We are [already filtering out reblogs](https://opengrok.a8c.com/source/xref/trunk/public.api/rest/wpcom-json-endpoints/class.wpcom-json-api-read-posts-related-endpoint.php#115). I think we will need to create a small project to fix this attribute when we index the data in Elasticsearch.

**Comment by blowery** [comment]
@samouri @bluefuton could you two give this a hairy 👀 ? I'd like to get it merged to fix LS.

File: client/lib/post-normalizer/rule-create-better-excerpt.js [hunk 3]
```diff
@@ -67,11 +68,12 @@ export default function createBetterExcerpt( post ) {
 	}
 
 	post.better_excerpt = formatExcerpt( post.content );
+	post.better_excerpt_no_html = stripHTML( post.better_excerpt );
 
 	// also make a shorter excerpt...
-	if ( post.excerpt ) {
+	if ( post.better_excerpt ) {
```
Comment (bluefuton) [inline client/lib/post-normalizer/rule-create-better-excerpt.js]: May as well just check for `post.better_excerpt_no_html` here?

**Comment by bluefuton** [comment]
Looks good!

Can we add some tests for the `removeElementsBySelector` rule before :shipit:?

**Comment by blowery** [comment]
@bluefuton tests added

**Pull request merged**